cpp/src/arrow/ipc/adapter.cc

```diff
@@ -0,0 +1,296 @@
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#include "arrow/ipc/adapter.h"
+
+#include <cstdint>
+#include <cstring>
+#include <vector>
+
+#include "arrow/array.h"
+#include "arrow/ipc/memory.h"
+#include "arrow/ipc/Message_generated.h"
+#include "arrow/ipc/metadata.h"
+#include "arrow/ipc/metadata-internal.h"
+#include "arrow/schema.h"
+#include "arrow/table.h"
+#include "arrow/type.h"
+#include "arrow/types/construct.h"
+#include "arrow/types/primitive.h"
+#include "arrow/util/buffer.h"
+#include "arrow/util/status.h"
+
+namespace arrow {
+namespace ipc {
+
+static bool IsPrimitive(const DataType* type) {
+  switch (type->type) {
+    case Type::NA:
+      return false;
+    case Type::BOOL:
+    case Type::UINT8:
+    case Type::INT8:
+    case Type::UINT16:
+    case Type::INT16:
+    case Type::UINT32:
+    case Type::INT32:
+    case Type::UINT64:
+    case Type::INT64:
+    case Type::FLOAT:
+    case Type::DOUBLE:
+      return true;
+    default:
+      return false;
+  }
+}
+
+// ----------------------------------------------------------------------
+// Row batch write path
+
+Status VisitArray(const Array* arr, std::vector<flatbuf::FieldNode>* field_nodes,
+    std::vector<std::shared_ptr<Buffer>>* buffers) {
+  if (IsPrimitive(arr->type().get())) {
+    const PrimitiveArray* prim_arr = static_cast<const PrimitiveArray*>(arr);
+
+    field_nodes->push_back(flatbuf::FieldNode(prim_arr->length(),
+            prim_arr->null_count()));
+
+    if (prim_arr->null_count() > 0) {
+      buffers->push_back(prim_arr->nulls());
+    } else {
+      // Push a dummy zero-length buffer, not to be copied
+      buffers->push_back(std::make_shared<Buffer>(nullptr, 0));
+    }
+    buffers->push_back(prim_arr->data());
+  } else if (arr->type_enum() == Type::LIST) {
+    // TODO(wesm)
+  } else if (arr->type_enum() == Type::STRUCT) {
+    // TODO(wesm)
+  }
+
+  return Status::OK();
+}
+
+class RowBatchWriter {
+ public:
+  explicit RowBatchWriter(const RowBatch* batch) :
+      batch_(batch) {}
+
+  Status AssemblePayload() {
+    // Perform depth-first traversal of the row-batch
+    for (int i = 0; i < batch_->num_columns(); ++i) {
+      const Array* arr = batch_->column(i).get();
+      RETURN_NOT_OK(VisitArray(arr, &field_nodes_, &buffers_));
+    }
+    return Status::OK();
+  }
+
+  Status Write(MemorySource* dst, int64_t position, int64_t* data_header_offset) {
+    // Write out all the buffers contiguously and compute the total size of the
+    // memory payload
+    int64_t offset = 0;
+    for (size_t i = 0; i < buffers_.size(); ++i) {
+      const Buffer* buffer = buffers_[i].get();
+      int64_t size = buffer->size();
+
+      // TODO(wesm): We currently have no notion of shared memory page id's,
```

Comment:
I'm not too familiar with the shared memory APIs, but I couldn't find reference to page ids directly via google.  The closest I could find might be some concepts from "shmget"?  But according shm_overview man page this doesn't seem like the right api to necessarily use (see below).   Is this an internal Arrow concept?  I would appreciate a pointer to the documentation

> "System  V  shared  memory (shmget(2), shmop(2), etc.) is an older shared memory API.  POSIX shared memory provides a simpler, and better
>        designed interface; on the other hand POSIX shared memory is somewhat less widely available (especially on older systems) than System  V
>        shared memory.
> "